libs/cardano-data/src/Data/UMap.hs

```diff
@@ -0,0 +1,691 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Data.UMap
+  ( Trip (Triple),
+    UMap (..),
+    VMap (..),
+    umInvariant,
+    unView,
+    unUnify,
+    rewView,
+    delView,
+    ptrView,
+    zero,
+    zeroMaybe,
+    mapNext,
+    mapLub,
+    next,
+    leastUpperBound,
+    empty,
+    delete,
+    delete',
+    insertWith,
+    insertWith',
+    insert,
+    insert',
+    lookup,
+    isNull,
+    domain,
+    range,
+    (∪),
+    (⨃),
+    (∪+),
+    (⋪),
+    (⋫),
+    member,
+    notMember,
+    domRestrict,
+    foldl',
+    Tag (..),
+    View (..),
+    findWithDefault,
+    size,
+    unify,
+  )
+where
+
+import Cardano.Binary (FromCBOR (..), ToCBOR (..), encodeListLen)
+import Control.DeepSeq (NFData (..))
+import Data.Coders (decodeMap, decodeRecordNamed, encodeMap)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.MapExtras (filterMaybe, intersectDomPLeft)
+import Data.Maybe.Strict (StrictMaybe (..))
+import Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Set.Internal as SI (Set (..))
+import Data.Sharing
+import Data.Typeable (Typeable)
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks (..))
+import Prelude hiding (lookup)
+
+-- ===================================================================
+
+{- The space compacting Trip datatype, and the pattern Triple are equivalent to:
+
+data Trip coin pool ptr = Triple
+  { coinT :: !(StrictMaybe coin),
+    ptrT :: !(Set ptr),
+    poolidT :: !(StrictMaybe pool)
+  }
+  deriving (Show, Eq, Generic, NoThunks, NFData)
+-}
+
+data Trip coin pool ptr
+  = TEEE
```

Comment:
`E` for empty, `F` for full?

Reply:
Yes Exactly, I added some comments to explain the convention.